tokio-io/src/split.rs

```diff
@@ -0,0 +1,175 @@
+//! Split a single value implementing `AsyncRead + AsyncWrite` into separate
+//! `AsyncRead` and `AsyncWrite` handles.
+//!
+//! To restore this read/write object from its `split::ReadHalf` and
+//! `split::WriteHalf` use `unsplit`.
+
+use crate::{AsyncRead, AsyncWrite};
+
+use bytes::{Buf, BufMut};
+use futures_core::ready;
+use std::cell::UnsafeCell;
+use std::fmt;
+use std::io;
+use std::pin::Pin;
+use std::sync::atomic::AtomicU8;
+use std::sync::atomic::Ordering::{Acquire, Release};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+/// The readable half of a value returned from `split`.
+pub struct ReadHalf<T> {
+    inner: Arc<Inner<T>>,
+}
+
+/// The writable half of a value returned from `split`.
+pub struct WriteHalf<T> {
+    inner: Arc<Inner<T>>,
+}
+
+struct Inner<T> {
+    state: AtomicU8,
+    stream: UnsafeCell<T>,
+}
+
+struct Guard<'a, T> {
+    inner: &'a Inner<T>,
+}
+
+const INIT: u8 = 0;
+const LOCKED: u8 = 1;
+
+/// Split a single value implementing `AsyncRead + AsyncWrite` into separate
+/// `AsyncRead` and `AsyncWrite` handles.
+///
+/// To restore this read/write object from its `split::ReadHalf` and
+/// `split::WriteHalf` use `unsplit`.
+pub fn split<T>(stream: T) -> (ReadHalf<T>, WriteHalf<T>) {
```

Comment:
I think we should add the `AsyncRead + AsyncWrite` here so that the compiler errors are nicer. I don't see anyone using this where they just have one half?

Reply:
Good idea.